src/debug_tools/diff.rs

```diff
@@ -0,0 +1,507 @@
+            DiffDetectionState::NoDiffDetected => 0,
+        }
+    }
+}
+
+/// A point (node) between the two tries where the children differ.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct DiffPoint {
+    pub depth: usize,
+    pub path: NodePath,
+    pub key: Nibbles,
+    pub a_info: NodeInfo,
+    pub b_info: NodeInfo,
+}
+
+impl DiffPoint {
+    fn new(
+        child_a: &HashedPartialTrie,
+        child_b: &HashedPartialTrie,
+        parent_k: Nibbles,
+        path: NodePath,
+    ) -> Self {
+        let a_key = parent_k.merge_nibbles(&get_key_piece_from_node_no_branch_key(child_a));
+        let b_key = parent_k.merge_nibbles(&get_key_piece_from_node_no_branch_key(child_b));
+
+        DiffPoint {
+            depth: 0,
+            path,
+            key: parent_k,
+            a_info: NodeInfo::new(child_a, a_key, get_value_from_node(child_a).cloned()),
+            b_info: NodeInfo::new(child_b, b_key, get_value_from_node(child_b).cloned()),
+        }
+    }
+}
+
+// TODO: Redo display method so this is more readable...
+impl Display for DiffPoint {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Point Diff {{depth: {}, ", self.depth)?;
+        write!(f, "Path: ({}), ", self.path)?;
+        write!(f, "Key: {:x} ", self.key)?;
+        write!(f, "A info: {} ", self.a_info)?;
+        write!(f, "B info: {}}}", self.b_info)
+    }
+}
+
+/// Meta information for a node in a trie.
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub struct NodeInfo {
+    key: Nibbles,
+
+    /// The direct value associated with the node (only applicable to `Leaf` &
+    /// `Branch` nodes).
+    value: Option<Vec<u8>>,
+    node_type: TrieNodeType,
+    hash: H256,
+}
+
+// TODO: Redo display method so this is more readable...
+impl Display for NodeInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "(key: {:x} ", self.key)?;
+
+        match &self.value {
+            Some(v) => write!(f, "Value: 0x{}, ", hex::encode(v))?,
+            None => write!(f, "Value: N/A, ")?,
+        }
+
+        write!(f, "Node type: {} ", self.node_type)?;
+        write!(f, "Trie hash: {:x})", self.hash)
+    }
+}
+
+impl NodeInfo {
+    fn new(n: &HashedPartialTrie, key: Nibbles, value: Option<Vec<u8>>) -> Self {
+        Self {
+            key,
+            value,
+            node_type: n.deref().into(),
+            hash: n.hash(),
+        }
+    }
+}
+
+/// Create a diff between two tries. Will perform both types of diff searches
+/// (top-down & bottom-up).
+pub fn create_diff_between_tries(a: &HashedPartialTrie, b: &HashedPartialTrie) -> TrieDiff {
+    TrieDiff {
+        latest_diff_res: find_latest_diff_point_where_tries_begin_to_diff(a, b),
+    }
+}
+
+// Only support `HashedPartialTrie` due to it being significantly faster to
+// detect differences due to hash caching.
+fn find_latest_diff_point_where_tries_begin_to_diff(
```

Comment:
nit: `due` repetition in doc.

This aside, I think we may want to rename it to something more compact.

Reply:
`find_latest_diff_point_between_tries`